### PR TITLE
RSA Keys + DERCodable + Cipher + Signature Conformance & Interoperability with Apple's Security Framework 

### PR DESCRIPTION
**Pull request by btoms20**
### This PR Introduces Additional RSA Key Functionality Including
- PKCS1v1.5 Encryption & Decryption
- PKCS1v1.5 Signatures and Signature Verification
- Raw Public & Private Key Exports & Imports (DER Format)
- Interoperability with Apple's `Security` Framework and RSA `SecKey`s 

### ‼️Due Diligence and Heads Up 
- This PR and the code it contains should be double & triple checked by the community before being approved. The more eyes the better. I do my best to write solid, clean, robust code, but I still make mistakes all of the time! So please, add more tests, especially for edge cases, ask questions, critique everything and lets make this code better together 🤝

-----
### Checklist:
- [x] Correct file headers (see CONTRIBUTING.md).
- [x] Formatted with [SwiftFormat](https://github.com/nicklockwood/SwiftFormat).
- [x] Tests added.
    - [PKCS1v1.5 Padding Tests](https://github.com/krzyzanowskim/CryptoSwift/compare/main...btoms20:CryptoSwift:feature/rsa+der?expand=1#diff-c238467c378d643ae033051e917376e71347aae172d48d4a8073b0a767be4eff)
    - [CryptoSwift <-> Security Interoperability Tests](https://github.com/krzyzanowskim/CryptoSwift/compare/main...btoms20:CryptoSwift:feature/rsa+der?expand=1#diff-946480c24138d055083a8ffef0fb4946df218b7fe388ace83009647c1ed33832)
    - [RSA Tests](https://github.com/krzyzanowskim/CryptoSwift/compare/main...btoms20:CryptoSwift:feature/rsa+der?expand=1#diff-11d478e123e946a6382a2c1852dc0ba89bc9400b67f1a90cd377c378bd3dd7d4)

-----
### 1) Introduced the ```Signature``` protocol ([Sources/CryptoSwift/Signature.swift](https://github.com/krzyzanowskim/CryptoSwift/compare/main...btoms20:CryptoSwift:feature/rsa+der?expand=1#diff-1857a0f739a3d18e1aab9cb742b7006e9812dc2148b2bc25838f16ae5c1dd8f2))
  - Asymmetric Key Pairs that support Signatures and Verification of said Signatures should conform to this protocol.
  - This protocol enforces the following two methods (and their ArraySlice counterparts).
    - ```func sign(_ bytes: Array<UInt8>) throws -> Array<UInt8> ```
    - ```func verify(signature: Array<UInt8>, for expectedData: Array<UInt8>) throws -> Bool```

-----
### 2) Introduced two PKCS1v15 Padding Schemes ([Sources/CryptoSwift/PKCS/PKCS1v15.swift](https://github.com/krzyzanowskim/CryptoSwift/compare/main...btoms20:CryptoSwift:feature/rsa+der?expand=1#diff-7879a5f4eb4769ba02d07efc30c173c1996d5e25a38dbafe397f3c7f174d96d8))
  - [EMSA-PKCS1v1_5 Padding Scheme IETF Spec](https://datatracker.ietf.org/doc/html/rfc8017#section-9.2)
    - EMSA is a **deterministic** padding scheme that is used during RSA Signature generation and Verification.
    - ```Encoded Message = 0x00 || 0x01 || PS || 0x00 || M``` (where `PS` is `0xFF` padding, and `M` is the original message). 
 - [EME-PKCS1v1_5 Padding Scheme IETF Spec](https://datatracker.ietf.org/doc/html/rfc8017#section-7.2.1)
    - EME is a **non-deterministic** padding scheme used during RSA Encryption and Decryption of messages.
    - ```Encoded Message = 0x00 || 0x02 || PS || 0x00 || M``` (where `PS` is pseudo-random, non zero bytes, and `M` is the original message).

-----
### 3) RSA DERCodable Conformance ([Sources/CryptoSwift/RSA/RSA.swift](https://github.com/krzyzanowskim/CryptoSwift/compare/main...btoms20:CryptoSwift:feature/rsa+der?expand=1#diff-e2773f8441924effb33c73a91e4d7980d0c34a9e9e0fb5f94868184dbf290c89))
  - RSA Keys now conform to `DEREncodable` and `DERDecodable` protocols (introduced in #922) which allow for the importing and exporting of RSA public and private keys in their "Raw External Representation".
  - This allows us to import RSA Keys that are generated using Apple's `Security` framework.
``` swift
/// Starting with a SecKey RSA Key
let rsaSecKey:SecKey

/// Copy External Representation
var externalRepError:Unmanaged<CFError>?
guard let cfdata = SecKeyCopyExternalRepresentation(rsaSecKey, &externalRepError) else {
  /// Failed to copy external representation for RSA SecKey
  return
}

/// Instantiate the RSA Key from the raw external representation
let rsaKey = try RSA(rawRepresentation: cfdata as Data)

/// You now have a CryptoSwift RSA Key
```

  - It also allows us to export CryptoSwift RSA keys for use with Apple's `Security` framework.

``` swift
/// Starting with a CryptoSwift RSA Key
let rsaKey = try RSA(keySize: 1024)

/// Define your Keys attributes
let attributes: [String:Any] = [
  kSecAttrKeyType as String: kSecAttrKeyTypeRSA,
  kSecAttrKeyClass as String: kSecAttrKeyClassPrivate, // or kSecAttrKeyClassPublic
  kSecAttrKeySizeInBits as String: 1024, // The appropriate bits
  kSecAttrIsPermanent as String: false
]
var error:Unmanaged<CFError>? = nil
guard let rsaSecKey = try SecKeyCreateWithData(rsaKey.externalRepresentation() as CFData, attributes as CFDictionary, &error) else {
  /// Error constructing SecKey from raw key data
  return
}

/// You now have an RSA SecKey for use with Apple's Security framework
```

-----
### 4) RSA Cipher Conformance ([Sources/CryptoSwift/RSA/RSA+Cipher.swift](https://github.com/krzyzanowskim/CryptoSwift/compare/main...btoms20:CryptoSwift:feature/rsa+der?expand=1#diff-a907d3bdb94042735c5018e3e634cca8ff9e1192d7db60834c57ff605dc15ba5))
- Added support for `.raw` and `.pkcs1v15` Encryption / Decryption variants.
- The previous Encryption / Decryption functionality is still available by using the `.unsafe` variant.
- These two new variants use the EME PKCS1 Padding Schemes (described above) to format data before & after encryption & decryption. 
- The functionality of both `.raw` and `.pkcs1v15` mimic that of Apple's `Security` framework
- The default variant used when calling `rsa.encrypt(...)` & `rsa.decrypt()` is `.pkcs1v15` due to it being non-deterministic and safer than it's deterministic alternatives.
``` swift
// Alice Generates a Private Key
let alicesPrivateKey = try RSA(keySize: 1024)
    
// Alice shares her **public** key with Bob
let alicesPublicKeyData = try alicesPrivateKey.publicKeyDER()
    
// Bob receives the raw external representation of Alices public key and imports it
let bobsImportOfAlicesPublicKey = try RSA(rawRepresentation: Data(alicesPublicKeyData))
    
// Bob can now encrypt a message for Alice using her public key
let message = "Hi Alice! This is Bob!"
let privateMessage = try bobsImportOfAlicesPublicKey.encrypt(message.bytes)
    
// This results in some encrypted output like this
// URcRwG6LfH63zOQf2w+HIllPri9Rb6hFlXbi/bh03zPl2MIIiSTjbAPqbVFmoF3RmDzFjIarIS7ZpT57a1F+OFOJjx50WYlng7dioKFS/rsuGHYnMn4csjCRF6TAqvRQcRnBueeINRRA8SLaLHX6sZuQkjIE5AoHJwgavmiv8PY=
      
// Bob can now send this encrypted message to Alice without worrying about people being able to read the original contents
    
// Alice receives the encrypted message and uses her private key to decrypt the data and recover the original message
let originalDecryptedMessage = try alicesPrivateKey.decrypt(privateMessage)
    
print(String(data: Data(originalDecryptedMessage), encoding: .utf8))
// "Hi Alice! This is Bob!"
```
- Decrypting data that was encrypted using Apple's `Security` framework is supported.

-----
### 5) RSA Signature Conformance ([Sources/CryptoSwift/RSA/RSA+Signature.swift](https://github.com/krzyzanowskim/CryptoSwift/compare/main...btoms20:CryptoSwift:feature/rsa+der?expand=1#diff-e1b27dfca23fa53c762d77bef0690fc02f55d9f0e112f296c7ceb62343908b8a))
- Added support for PKCS1v15 `.message` and `.digest` varieties of RSA Signatures
- The `.message` variants hash the data before signing it using the specified Hash function.
- The `.digest` variants expects a pre-hashed message digest (therefore skips the hashing of the data) before signing.
- This behavior, and naming convention, is consistent with Apple's `Security` framework.
- The default signature variant is `.message_pkcs1v15_SHA256`
- The supported hash functions are `MD5`, `SHA1`, `SHA224`, `SHA256`,  `SHA384`,  `SHA512`, `SHA512_224` and `SHA512_256`
``` swift
// Alice Generates a Private Key
let alicesPrivateKey = try RSA(keySize: 1024)
    
// Alice wants to sign a message that she agrees with
let messageAliceSupports = "Hi my name is Alice!"
let alicesSignature = try alicesPrivateKey.sign(messageAliceSupports.bytes)
    
// Alice shares her Public key and the signature with Bob
let alicesPublicKeyData = try alicesPrivateKey.publicKeyDER()
    
// Bob receives the raw external representation of Alices Public key and imports it!
let bobsImportOfAlicesPublicKey = try RSA(rawRepresentation: Data(alicesPublicKeyData))
        
// Bob can now verify that Alice signed the message using the Private key associated with her shared Public key.
let verifiedSignature = try bobsImportOfAlicesPublicKey.verify(signature: alicesSignature, for: "Hi my name is Alice!".bytes)
    
if verifiedSignature == true {
  // Bob knows that the signature Alice provided is valid for the message and was signed using the Private key associated with Alices shared Public key.
} else {
  // The signature was invalid, so either
  // - the message Alice signed was different then what we expected.
  // - or Alice used a Private key that isn't associated with the shared Public key that Bob has.
}
```
-----
### RSA Test Fixtures
The RSA Key Test Fixtures located in `Tests/CryptoSwiftTests/RSATests.swift` were created using Apple's `Security` Framework. More fixtures can be created by executing the `testCreateTestFixture` test in `Tests/CryptoSwiftTests/RSASecKeyTests.swift`. Simply adjust the key size and the messages you'd like to sign / encrypt and a new Fixture object will be logged to the console. These fixtures are extremely helpful when testing against deterministic signature and encryption schemes since we can compare the data at both stages. But they are also helpful for decryption and verification of non-deterministic encryption and signature schemes respectively. Hopefully using one cryptography library to test another is seen as acceptable.

### CODE REVIEW

**Comment by NathanFallet** [comment]
I'm going to take a look and review everything carefully as well (other reviews are welcomed too)

**Review by NathanFallet — COMMENTED** [review]
I need to check tests again because I miss of time for now, but for the rest I added some comments.

File: Sources/CryptoSwift/RSA/RSA.swift [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+    guard case .integer(let publicExponent) = params[2] else { throw Error.invalidDERFormat }
+    guard case .integer(let privateExponent) = params[3] else { throw Error.invalidDERFormat }
+    guard case .integer(let prime1) = params[4] else { throw Error.invalidDERFormat }
+    guard case .integer(let prime2) = params[5] else { throw Error.invalidDERFormat }
+    guard case .integer(let exponent1) = params[6] else { throw Error.invalidDERFormat }
+    guard case .integer(let exponent2) = params[7] else { throw Error.invalidDERFormat }
+    guard case .integer(let coefficient) = params[8] else { throw Error.invalidDERFormat }
+
+    // Are there other versions out there? Is this even the version?
+    guard version == Data(hex: "0x00") else { throw Error.invalidDERFormat }
```
[the remark this answers is not in the document]
Comment (btoms20) [reply Sources/CryptoSwift/RSA/RSA.swift:199]: Yeah, my apologies, I grabbed the sequence definition from [RFC3447](https://www.rfc-editor.org/rfc/rfc3447#appendix-A.1.2) (which added support for RSA Keys with multiple primes). The [RFC2313](https://datatracker.ietf.org/doc/html/rfc2313#section-7.2) private key syntax doesn't include the optional `otherPrimeInfos`. And RFC2313 is the only version tested and supported in this PR. If we’d like to add support for RFC3447 then we can add that functionality later in a different PR?

I'll update the comment to be more informative and add a proper error case.

``` swift
// We only support version 0x00 == RFC2313 at the moment
// - TODO: Support multiple primes 0x01 version defined in [RFC3447](https://www.rfc-editor.org/rfc/rfc3447#appendix-A.1.2)
guard version == Data(hex: "0x00") else { throw Error.unsupportedRSAVersion }
```

File: Sources/CryptoSwift/RSA/RSA.swift [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+  ///   coefficient       INTEGER,  -- (inverse of q) mod p
+  ///   otherPrimeInfos   OtherPrimeInfos OPTIONAL
+  /// }
+  /// ```
+  internal convenience init(privateDER der: Array<UInt8>) throws {
+    let asn = try ASN1.Decoder.decode(data: Data(der))
+
+    // Enforce the above ASN Structure (do we need to extract and verify the eponents and coefficients?)
+    guard case .sequence(let params) = asn else { throw Error.invalidDERFormat }
+    guard params.count >= 9 else { throw Error.invalidDERFormat }
```
[the remark this answers is not in the document]
Comment (btoms20) [reply Sources/CryptoSwift/RSA/RSA.swift:187]: As defined in RFC2313 there has to be exactly 9 parameters. I was conflating RFC3447 and RFC2313.

I'll update the code as follows...
``` swift
guard params.count == 9 else { throw Error.invalidDERFormat }
```

File: Sources/CryptoSwift/RSA/RSA.swift [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+    // Calculate modulus
+    guard BigUInteger(modulus) == BigUInteger(prime1) * BigUInteger(prime2) else { throw Error.invalidPrimes }
+
+    // Calculate public and private exponent
+    let phi = (BigUInteger(prime1) - 1) * (BigUInteger(prime2) - 1)
+    guard let d = BigUInteger(publicExponent).inverse(phi) else { throw Error.invalidPrimes }
+    guard BigUInteger(privateExponent) == d else { throw Error.invalidPrimes }
+
+    // Ensure the provided coefficient is correct (derived from the primes)
+    // - Note: this might be overkill, cause we don't store the coefficient, but the extra check probably isn't the worse thing
```
[the remark this answers is not in the document]
Comment (btoms20) [reply Sources/CryptoSwift/RSA/RSA.swift:211]: Agreed! 👍

File: Sources/CryptoSwift/RSA/RSA.swift [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+  ///   version           Version,
+  ///   modulus           INTEGER,  -- n
+  ///   publicExponent    INTEGER,  -- e
+  ///   privateExponent   INTEGER,  -- d
+  ///   prime1            INTEGER,  -- p
+  ///   prime2            INTEGER,  -- q
+  ///   exponent1         INTEGER,  -- d mod (p-1)
+  ///   exponent2         INTEGER,  -- d mod (q-1)
+  ///   coefficient       INTEGER,  -- (inverse of q) mod p
+  ///   otherPrimeInfos   OtherPrimeInfos OPTIONAL
```
[the remark this answers is not in the document]
Comment (btoms20) [reply Sources/CryptoSwift/RSA/RSA.swift:269]: I was conflating RFC3447 and RFC2313. But to be clear, the only version supported and tested in this PR is RFC2313. I'll be removing the `otherPrimeInfos` optional from the code and comments.

File: Sources/CryptoSwift/RSA/RSA.swift [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+// Foundation is required for `Data` to be found
+import Foundation
+
+// Note: The `BigUInt` struct was copied from:
+// https://github.com/attaswift/BigInt
+// It allows fast calculation for RSA big numbers
+
+public final class RSA: DERCodable {
+  /// RSA Object Identifier Bytes (rsaEncryption)
+  static var primaryObjectIdentifier: Array<UInt8> = Array<UInt8>(arrayLiteral: 0x2A, 0x86, 0x48, 0x86, 0xF7, 0x0D, 0x01, 0x01, 0x01)
```
[the remark this answers is not in the document]
Comment (btoms20) [reply Sources/CryptoSwift/RSA/RSA.swift:25]: Ah, thanks for pointing that out. I was working on PEM support simultaneously	 and the `primary` and `secondary` PKCS#8 OID’s are used when constructing a PEM formatted representation of the key. Although it’s not necessary for the Raw External DER Representation of a key, If the DERCodable protocol enforces these OID’s then we can extend the protocol with PEM functionality. Here’s an example of what I had in mind…
``` swift
public extension DEREncodable {

  internal func exportPublicKeyPEMRaw() throws -> Array<UInt8> {
    let publicDER = try self.publicKeyDER()

    let node:ASN1.Node = .sequence(nodes: [
        .sequence(nodes: [
          .objectIdentifier(data: Data(Self.primaryObjectIdentifier)),
          .null
        ]),
        .bitString(data: Data( publicDER ))
      ])
      
    return ASN1.Encoder.encode(node)
  }

  func exportPublicKeyPEM(withHeaderAndFooter:Bool = true) throws -> Array<UInt8> {
    let base64String = try self.exportPublicKeyPEMRaw().toBase64()
    let bodyString = base64String.chunks(ofCount: 64).joined(separator: "\n")
    let bodyUTF8Bytes = bodyString.bytes
    
    if withHeaderAndFooter {
      let header = PEM.PEMType.publicKey.headerBytes + [0x0a]
      let footer = [0x0a] + PEM.PEMType.publicKey.footerBytes
    
      return header + bodyUTF8Bytes + footer
    } else {
      return bodyUTF8Bytes
    }
  }
```

But maybe it makes more sense to define a PEM protocol in the future that extends DERCodable and enforces the PKCS8 OID’s. Something like…
``` swift
protocol PEM:DERCodable {
  /// The keys ASN1 primary object identifier (ex: RSA --> rsaEncryption --> [0x2A, 0x86, 0x48, 0x86, 0xF7, 0x0D, 0x01, 0x01, 0x01])
  static var primaryObjectIdentifier:Array<UInt8> { get }
  /// The keys ASN1 secondary object identifier (ex: RSA --> .null)
  static var secondaryObjectIdentifier:Array<UInt8>? { get }
}

extension PEM {
  /// Instantiates a DERDecodable Key from a PEM string
  /// - Parameters:
  ///   - pem: The PEM file to import
  ///   - password: A password to use to decrypt an encrypted PEM file
  ///   - asType: The underlying DERDecodable Key Type (ex: RSA.self)
  init<Key:DERDecodable>(pem: String, password: String? = nil, asType:Key.Type = Key.self) throws {
      …
  }

  /// PublicKey PEM Export Functions
  func exportPublicKeyPEM(withHeaderAndFooter:Bool) throws -> Array<UInt8> { … }
  func exportPublicKeyPEMString(withHeaderAndFooter:Bool) throws -> String { … }
  
  /// PrivateKey PEM Export Functions
  func exportPrivateKeyPEM(withHeaderAndFooter:Bool) throws -> Array<UInt8> { … }
  func exportPrivateKeyPEMString(withHeaderAndFooter:Bool) throws -> String { .. }
}
```
If you’d prefer the later, then I can go ahead and remove the `primary` and `secondary` OID’s from `RSA` and both of the `DEREncodable` and `DERDecodable` protocols.
 
Let me know which you prefer (I think I prefer the later, especially since there might be DERCodable objects that don't require / have a primary and secondary object id).

File: Sources/CryptoSwift/RSA/RSA+Cipher.swift [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    // Check for Private Exponent presence
+    guard let d = d else { throw RSA.Error.noPrivateKey }
+
+    // Calculate decrypted data
+    return BigUInteger(Data(bytes)).power(d, modulus: self.n).serialize().bytes
+  }
+}
+
+extension RSA {
+  public enum RSAEncryptionVariant {
```
[the remark this answers is not in the document]
Comment (btoms20) [reply Sources/CryptoSwift/RSA/RSA+Cipher.swift:70]: I was working off of [RFC2313 Section 8.1 - Encryption block formatting](https://datatracker.ietf.org/doc/html/rfc2313#section-8.1), and taking into account Apple's `Security` framework. `Security` uses raw (zero padded, block type 0) and pkcs1v15 (pseudo random, block type 2) for encryption (public key operations) and reserves (255 padded, block type 1) for Signatures (private key operation). I tried to match that functionality.

File: Sources/CryptoSwift/RSA/RSA.swift [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+    }
+  }
+}
+
+// MARK: CS.BigUInt extension
+
+extension BigUInteger {
+
+  public static func generatePrime(_ width: Int) -> BigUInteger {
+    // Note: Need to find a better way to generate prime numbers
```
[the remark this answers is not in the document]
Comment (btoms20) [reply Sources/CryptoSwift/RSA/RSA.swift:361]: I was actually playing around with multi threaded prime number searches this morning and measuring the change in perfomance. It uses the same exact code as before, random guessing and checking, just does it in parallel when possible. Here's some really rough code that compiles on both Mac and Linux, although I don't have a multi-core linux machine to test it on at the moment...

``` swift
internal class MultithreadedPrimeGeneration {
    let numberOfPrimes:Int
    let size:Int
    let threadCount:Int
    
    private var group:[Thread] = []
    private var isDone:Bool = false
    private var primesGenerated:[BigUInteger] {
      didSet {
        if primesGenerated.count >= numberOfPrimes {
          self.isDone = true
          self.group.forEach { $0.cancel() }
          self.group = []
        }
      }
    }
    
    init(numberOfPrimes:Int = 2, size:Int, threadCount:Int = 1) {
      self.numberOfPrimes = numberOfPrimes
      self.size = size
      self.threadCount = threadCount
      self.primesGenerated = []
    }
    
    public func generate() -> [BigUInteger] {
      guard threadCount != 1 else {
        print("Single Thread")
        return (0..<numberOfPrimes).map { _ in generatePrime(self.size) }
      }
      print("Launching [\(threadCount)] Threads in search of Primes!")
      self.startMulti()
      while !isDone { usleep(100_000) }
      return Array(primesGenerated.prefix(numberOfPrimes))
    }
    
    public func startMulti() {
      group = (0..<threadCount).map { _ in
        Thread {
          while !self.isDone {
            if let prime = self.searchForPrime(self.size) { self.primesGenerated.append(prime) }
          }
        }
      }
      group.forEach { $0.start() }
    }
    
    /// Searches for a prime as long as isDone is false
    private func searchForPrime(_ width: Int) -> BigUInteger? {
      while !self.isDone {
        var random = BigUInteger.randomInteger(withExactWidth: width)
        random |= BigUInteger(1)
        if random.isPrime() {
          return random
        }
      }
      return nil
    }
    
    /// Searches for a prime until it finds one
    private func generatePrime(_ width: Int) -> BigUInteger {
      while true {
        var random = BigUInteger.randomInteger(withExactWidth: width)
        random |= BigUInteger(1)
        if random.isPrime() {
          return random
        }
      }
    }
  }
```

and you'd use it like so...
``` swift
/// Initialize with a generated key pair
/// - Parameter keySize: The size of the modulus
public convenience init(keySize: Int, threads:Int = 1) throws {
  // Generate prime numbers
  //let p = BigUInteger.generatePrime(keySize / 2)
  //let q = BigUInteger.generatePrime(keySize / 2)
  let primes = MultithreadedPrimeGeneration(numberOfPrimes: 2, size: keySize / 2, threadCount: threads).generate()
  let p = primes.first!
  let q = primes.last!
  
  // Calculate modulus
  let n = p * q

  // Calculate public and private exponent
  let e: BigUInteger = 65537
  let phi = (p - 1) * (q - 1)
  guard let d = e.inverse(phi) else {
    throw RSA.Error.invalidInverseNotCoprimes
  }

  // Initialize
  self.init(n: n, e: e, d: d, p: p, q: q)
}
```

Key generation times are listed below in **milliseconds**.
- Note: This is the time it takes to find 2 primes of keySize / 2
- Note: The code was built in production mode using the `--configuration release` flag
- Note: You can detect how many cores a Mac has with a call to `sysconf(CInt(_SC_NPROCESSORS_ONLN))`. I'm not sure if this works on Linux (I would assume it wouldn't)

| Threads | 1024 | 2048 | 3072  | 4096  |
|---------|------|------|-------|-------|
| 1       | 247  | 2644 | 11472 | 19081 |
| 2       | 452  | 1827 | 6379  | 20427 |
| 4       | 390  | 1153 | 3488  | 7974  |
| 8       | 391  | 808  | 2657  | 7479  |


I also came across this repo that played around with using Metal shaders to generate primes.
https://github.com/dneprDroid/PrimeNumbers-Metal
But I haven't had a chance to try it out yet, and I don't think it's a great cross platform solution.

File: Sources/CryptoSwift/RSA/RSA.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (NathanFallet) [reply Sources/CryptoSwift/RSA/RSA.swift:25]: I think that the second approach is better, because not all DERCodable objects require PEM support. And also I think that `PEMCodable` would be more consistant for the name of the protocol.

File: Sources/CryptoSwift/RSA/RSA.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (NathanFallet) [reply Sources/CryptoSwift/RSA/RSA.swift:361]: That's already an improvement! Also see what we talked about in #933

**Review by NathanFallet — APPROVED** [review]
It looks good to me.
For PEM related protocol, do you plan to make another PR or include it in this one? And also if you find more things about prime generation, don't hesitate to open a PR for it.

**Comment by btoms20** [comment]
> For PEM related protocol, do you plan to make another PR or include it in this one?

PEM support should probably be another PR.

>  if you find more things about prime generation, don't hesitate to open a PR for it.

Will do!

**Comment by NathanFallet** [comment]
Great! I merged your PR, and I will be here to review any other PR you open
